espree.js

```diff
@@ -64,18 +63,94 @@ function resetExtra() {
 
 var tt = acorn.tokTypes,
     Parser = acorn.Parser,
-    pp = Parser.prototype;
+    pp = Parser.prototype,
+    getLineInfo = acorn.getLineInfo;
 
 
-// hack Node
+function isValidNode(node) {
+    var ecma = extra.ecmaFeatures;
 
-var finishNode = pp.finishNode,
-    finishNodeAt = pp.finishNodeAt,
-    eat = pp.eat;
+    switch (node.type) {
+        case "VariableDeclaration":
+            return node.kind === "var" || ecma.blockBindings;
+
+        case "ObjectPattern":
+        case "ArrayPattern":
+            return ecma.destructuring;
+
+        case "AssignmentPattern":
+            // TODO: enhance analysis for separate options
+            return ecma.destructuring || ecma.defaultParams;
+
+        case "RestElement":
+            // TODO: enhance analysis for separate options
+            return ecma.destructuring || ecma.restParams;
+
+        case "ForOfStatement":
+            return ecma.forOf;
+
+        case "Property":
+            if (node.computed && !ecma.objectLiteralComputedProperties) {
+                return false;
+            }
+            if (node.method && !ecma.objectLiteralShorthandMethods) {
+                return false;
+            }
+            if (node.shorthand && !ecma.objectLiteralShorthandProperties) {
+                return false;
+            }
+            // TODO: analyse for objectLiteralDuplicateProperties: false in ES6
+            return true;
+
+        case "FunctionExpression":
+        case "FunctionDeclaration":
+            if (node.generator && !ecma.generators) {
+                return false;
+            }
+            return true;
+
+        case "YieldExpression":
+            return ecma.generators;
+
+        case "SpreadElement":
+            return ecma.spread;
+
+        case "ClassDeclaration":
+        case "ClassExpression":
+            return ecma.classes;
+
+        case "Super":
+            // TODO: enhance analysis for separate options
+            return ecma.classes || ecma.superInFunctions;
+
+        case "ImportDeclaration":
+        case "ExportNamedDeclaration":
+        case "ExportDefaultDeclaration":
+        case "ExportAllDeclaration":
+            return ecma.modules;
+
+        default:
+            return true;
+    }
+}
 
 function esprimaFinishNode(result) {
-    delete result.start;
-    delete result.end;
+    // ensure that parsed node was allowed through ecmaFeatures
+    if (!isValidNode(result)) {
+        this.unexpected(result.start);
+    }
+
+    // TODO: remove when marijnh/acorn#281 is resolved
+    if (result.type === "AssignmentPattern") {
+        delete result.operator;
+    }
+
+    // hide acorn-specific properties from comparison
+    // but leave for internal needs:
+    Object.defineProperties(result, {
```

Comment:
Not really because Acorn and Acorn-JSX heavily rely on them for internal needs since instead of look-aheads they often use look-behind technique which means looking back at some already parsed nodes and talking some info, incl. offsets, from them.